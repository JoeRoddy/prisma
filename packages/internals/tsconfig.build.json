{
  "extends": "../../tsconfig.build.bundle.json",
  "compilerOptions": {
    "outDir": "declaration",
    "typeRoots": ["./node_modules/@types", "./typings"]
  },
  "include": ["src"]
}
