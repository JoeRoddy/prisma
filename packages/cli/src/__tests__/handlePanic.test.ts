import { jestConsoleContext, jestContext } from '@prisma/get-platform'
import { canPrompt, handlePanic } from '@prisma/internals'
import { DbPull, MigrateDev } from '@prisma/migrate'

const ctx = jestContext.new().add(jestConsoleContext()).assemble()

describe('handlePanic', () => {
  // testing with env https://stackoverflow.com/a/48042799/1345244
  const OLD_ENV = process.env

  // mock for retrieving the database version
  const getDatabaseVersionSafe = () => Promise.resolve(undefined)

  beforeEach(() => {
    jest.resetModules() // most important - it clears the cache
    process.env = { ...OLD_ENV } // make a copy
  })
  afterAll(() => {
    process.env = OLD_ENV // restore old env
  })

  const packageJsonVersion = '0.0.0'
  const enginesVersion = '734ab53bd8e2cadf18b8b71cb53bf2d2bed46517'
  const command = 'something-test'

  describe('migrate', () => {
    it('no interactive mode in CI', async () => {
      expect.assertions(6)
      process.env.FORCE_PANIC_MIGRATION_ENGINE = '1'
      ctx.fixture('migration-engine-interactive')

      const migrate = new MigrateDev()

      let error: Error | undefined

      try {
        await migrate.parse(['--name', 'setup'])
      } catch (e) {
        const rustPanic = e

        expect(JSON.parse(JSON.stringify(rustPanic))).toMatchObject({
          area: 'LIFT_CLI',
        })
        expect(rustPanic).toMatchSnapshot()
        expect(rustPanic.message).toContain('This is the debugPanic artificial panic')
        expect(rustPanic.rustStack).toContain('[EXIT_PANIC]')

        const isWindows = ['win32'].includes(process.platform)
        if (!isWindows) {
          expect(rustPanic.rustStack).toContain('std::panicking::')
        }

        try {
          await handlePanic({
            error: rustPanic,
            cliVersion: packageJsonVersion,
            enginesVersion,
            command,
            getDatabaseVersionSafe,
          })
        } catch (err) {
          error = err
        }
      }

      if (!canPrompt()) {
        expect(error).toMatchInlineSnapshot(`
          Error in migration engine.
          Reason: [/some/rust/path:0:0] This is the debugPanic artificial panic

        `)
      }
    })
  })

  describe('introspect', () => {
    it('no interactive mode in CI', async () => {
      expect.assertions(6)
      process.env.FORCE_PANIC_MIGRATION_ENGINE = '1'
      ctx.fixture('migration-engine-interactive')

      const migrate = new DbPull()

      let error: Error | undefined

      try {
        await migrate.parse(['--print'])
      } catch (e) {
        const rustPanic = e

        expect(JSON.parse(JSON.stringify(rustPanic))).toMatchObject({
          area: 'LIFT_CLI',
        })
        expect(rustPanic).toMatchSnapshot()
        expect(rustPanic.message).toContain('This is the debugPanic artificial panic')
        expect(rustPanic.rustStack).toContain('[EXIT_PANIC]')

        const isWindows = ['win32'].includes(process.platform)
        if (!isWindows) {
          expect(rustPanic.rustStack).toContain('std::panicking::')
        }

        try {
          await handlePanic({
            error: rustPanic,
            cliVersion: packageJsonVersion,
            enginesVersion,
            command,
            getDatabaseVersionSafe,
          })
        } catch (err) {
          error = err
        }
      }

      if (!canPrompt()) {
        expect(error).toMatchInlineSnapshot(`
          Error in migration engine.
          Reason: [/some/rust/path:0:0] This is the debugPanic artificial panic

        `)
      }
    })
  })
})
