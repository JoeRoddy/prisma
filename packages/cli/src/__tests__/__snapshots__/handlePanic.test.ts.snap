// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlePanic introspect no interactive mode in CI 1`] = `
Error in migration engine.
Reason: [/some/rust/path:0:0] This is the debugPanic artificial panic

`;

exports[`handlePanic migrate no interactive mode in CI 1`] = `
Error in migration engine.
Reason: [/some/rust/path:0:0] This is the debugPanic artificial panic

`;
